packages/design-system/src/components/date-input/date-input.tsx

```diff
@@ -0,0 +1,218 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under
+ * the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
+ * OF ANY KIND, either express or implied. See the License for the specific language
+ * governing permissions and limitations under the License.
+ */
+
+import {
+  type ForwardedRef,
+  Fragment,
+  createContext,
+  forwardRef,
+  useCallback,
+  useContext,
+  useMemo,
+} from 'react';
+import {
+  type ContextValue,
+  DateFieldStateContext,
+  Group,
+  DateSegment as RACDateSegment,
+  type SlotProps,
+  TimeFieldStateContext,
+} from 'react-aria-components';
+import { useContextProps, useDefaultProps, useTheme } from '../../hooks';
+import { inputs } from '../../styles';
+import { callRenderProps, inlineVars, mergeClassNames } from '../../utils';
+import { Input } from '../input';
+import {
+  dateInputClassNames,
+  dateInputStateVars,
+  dateSegmentStateVars,
+} from './date-input.css';
+import type {
+  DateInputProps,
+  DateInputRenderProps,
+  DateSegmentProps,
+  DateSegmentRenderProps,
+  DateSegmentsProps,
+} from './types';
+
+const defaultMapping = {
+  icon: {
+    sm: { size: 'xs' },
+    lg: { size: 'md' },
+  },
+  input: {
+    sm: inputs.sm,
+    lg: inputs.lg,
+  },
+};
+
+const defaultSize = 'lg';
+
+export const DateInputContext =
+  createContext<ContextValue<DateInputProps, HTMLDivElement>>(null);
+
+export const DateInput = forwardRef(function DateInput(
+  props: DateInputProps,
+  ref: ForwardedRef<HTMLDivElement>,
+) {
+  [props, ref] = useContextProps(props, ref, DateInputContext);
+  props = useDefaultProps(props, 'DateInput');
+
+  const {
+    children: childrenProp,
+    classNames: classNamesProp,
+    mapping: mappingProp,
+    size = defaultSize,
+    provider,
+    ...rest
+  } = props;
+
+  const dateFieldState = useContext(DateFieldStateContext);
+  const timeFieldState = useContext(TimeFieldStateContext);
+  const state = dateFieldState ?? timeFieldState ?? null;
+
+  const theme = useTheme();
+
+  const mapping = useMemo(
+    () => ({
+      ...defaultMapping,
+      ...mappingProp,
+    }),
+    [mappingProp],
+  );
+
+  const classNames = useMemo(
+    () =>
+      mergeClassNames(dateInputClassNames, theme.DateInput, classNamesProp, {
+        input: { input: mapping.input[size] },
+      }),
+    [theme.DateInput, classNamesProp, mapping, size],
+  );
+
+  const style = useCallback(
+    (renderProps: DateInputRenderProps) =>
+      inlineVars(dateInputStateVars, { ...renderProps, size }),
+    [size],
+  );
+
+  const children = useCallback(
+    (renderProps: DateInputRenderProps) => (
+      <div className={classNames?.input?.input}>
+        {childrenProp &&
+          (provider ? (
+            callRenderProps(childrenProp, { ...renderProps, ...state })
+          ) : (
+            <>
+              {state.segments.map((segment, i) => (
+                <Fragment key={i}>{childrenProp(segment)}</Fragment>
```

Comment:
We could run into issues using the array index as a key, maybe something like this?
```suggestion
                <Fragment key={`${i}_${segment.text}`}>{childrenProp(segment)}</Fragment>
```

Reply:
actually, I just realized I can use the type of segment -- year, day, etc -- since there should be no duplicates.